### PR TITLE
Better sizehint for primes

### PR DESCRIPTION
**Pull request by mschauer**
Better sizehint for `primes(lo,hi)`.

Before it failed for 1 << lo < hi as the sizehint! actually consumed  `hi / log(hi)` memory independent of `lo`, but the needed size is better approximated by  `hi / log(hi) - lo/log(lo)`. Add a test which test `primes` and friends on short intervals with `1 << lo`.

### CODE REVIEW

File: test/numbers.jl [hunk 1]
```diff
@@ -2253,6 +2253,9 @@ for T in [Int,BigInt], n = [1:1000;1000000]
         @test s[k] == primesmask(k, k)[1]
     end
 end
+let i = rand(1:2^40)
```
Comment (tkelman) [inline test/numbers.jl]: `2^40` overflows on 32 bit

Comment (mschauer) [inline test/numbers.jl]: Oh, yes, and I swear I did not just try to make a point here - #5573

File: base/primes.jl [hunk 2]
```diff
@@ -83,7 +83,7 @@ function primes(lo::Int, hi::Int)
     lo ≤ 3 ≤ hi && push!(list, 3)
     lo ≤ 5 ≤ hi && push!(list, 5)
     hi < 7 && return list
-    sizehint!(list, floor(Int, hi / log(hi)))
+    sizehint!(list, floor(Int, max(hi,0) / log(max(hi,2)) - max(lo,0) / log(max(lo,2))))
```
Comment (pabloferz) [inline base/primes.jl]: I think you don't need `max` here except for `max(lo,0)` and `max(lo,2)` as `hi` will be greater or equal than 7 at this point.

Comment (pabloferz) [inline base/primes.jl]: I guess we can clean this up a bit by writing `if lo < 2; lo = 2; end` after the `lo ≤ hi` check. And then just write `hi / log(hi) - lo / log(lo)` for the sizehint

**Comment by mschauer** [comment]
Thanks for the feedback. Not a big surprise, but for example now the batched infinite prime iterator 
``
    primeiter = Base.flatten(primes(1000000*(i-1)+1,1000000*i) for i in countfrom(1))
``
actually works 
<pre>
julia-before> @time first(drop(primeiter, 10^9))
 93.870623 seconds (2.00 G allocations: <b>81.113 TB</b>, 10.60% gc time)
22801763513

julia-after> @time first(drop(primeiter, 10^9))
 88.758689 seconds (2.00 G allocations: <b>87.218 GB</b>, 3.50% gc time)
22801763513
</pre>

For comparison, 
<pre>
julia> @time primes(22801763513)[end];
114.720337 seconds (10 allocations: 13.415 GB, 0.15% gc time)
</pre>
takes a bit longer.

**Comment by pabloferz** [comment]
We could use better bound than `n/log(n)`, but if the increase in time for big numbers is not that bad this PR might be a good compromise.

**Comment by mschauer** [comment]
It's only slightly too big (with relative error of 5% for n=10^9 for example), which is good. That is better than choosing something too small but much closer (Li(n) for example).

**Comment by pabloferz** [comment]
A much better bound for `n>3` is `n/(log(n)-1.12)`. Se here https://projecteuclid.org/download/pdf_1/euclid.rmjm/1181070157

**Comment by pabloferz** [comment]
But that will only work for `n > 8` as it is a decreasing function before that point.

**Comment by pabloferz** [comment]
Maybe something like

```julia
π_hi = hi / (log(hi) - (hi > 7 ? 1.12 : 0))
π_lo = lo / (log(lo) - (lo > 7 ? 1.12 : 0))
sizehint!(list, floor(Int, π_hi - π_lo))
```

**Comment by pabloferz** [comment]
Ok, as we have

`x / log(x) < π(x) < x / (log(x) - 1.12)`

then

`hi  / log(hi) - lo / (log(lo) - 1.12) < π_hi - π_lo < hi / (log(hi) - 1.12) - lo / log(lo)`

so it would probably be better to have (though it would use more memory than you have now.

`sizehint!(list, floor(Int, hi / (log(hi) - 1.12) - lo / log(lo))`

I guess is just a matter of experimenting a bit.

**Comment by mschauer** [comment]
Do we have `` π(x) < x / (log(x) - 1.12)`` ?

**Comment by pabloferz** [comment]
See equation 10 from the reference a pasted above

**Comment by mschauer** [comment]
``hi / (log(hi) - 1.12) - lo / log(lo)`` is much too big if hi-lo is relatively small. I'll take 
``hi / (log(hi) - 1.12) - lo / (log(lo) - 1.12)`` to make it a tighter upper bound in the standard setting ``primes(n)``. That also shaves another 10 percent of the memory consumption in the example.

**Comment by pabloferz** [comment]
Just beware of `lo / (log(lo) - 1.12)` when `lo < 4` you should use `lo / (log(lo) - (lo > 3 ? 1.12 : 0)` or something like that

**Comment by mschauer** [comment]
I think we can take that one. That should work fine in practice, as sizehint!() round to the next multiple of a resizing factor and does not need much precision, only good overall estimates which give enough space to the array.

**Comment by pabloferz** [comment]
LGTM.

**Comment by mschauer** [comment]
@ararslan Do you want to have a look? I think this is good to go.

**Comment by ararslan** [comment]
@mschauer Funny you should ask me, I'm just some guy. :stuck_out_tongue_winking_eye: 

Looks good to me (and it appears the AppVeyor failure was just a timeout). What, if any, are the impacts on performance? Do you know?

**Comment by ararslan** [comment]
Want to try closing and reopening the PR to see if we can get AppVeyor to cooperate?

**Comment by mschauer** [comment]
Thanks to the unknown helping hand triggering AppVeyor.

**Comment by mschauer** [comment]
#16357

**Pull request closed without merging**